Add ezSimdDouble class

TrevorCash opened the pull request:
Adds ezSimdDouble class with AVX implementation.

Tested on x86_64 Linux with a corresponding unit test. 

Needs feedback on the Sqrt and Recipricol Related correctness.

TrevorCash committed:
Add ezSimdDouble class.  add AVX implementation.

C-Core commented:
To be honest, I don't see the point in why we should have a simd double implementation. It is not something commonly used in games. And even if we want to add this, it should be an SSE level, (we are currently on SSE 4.1, so something like SSE AVX) and not orthogonal to SSE. Also you cannot simply set AVX compiler flags because that will make AVX a minimum requirement which means it will crash on a non-AVX cpu.

jankrassnigg commented:
I do think that having a double version could be useful, since we may want to support double precision for game object positions at some point. So having this is generally fine IMO. But I agree that we can't just enable the AVX requirement, that would break the engine for many people.

TrevorCash committed:
Add EZ_REQUESTED_SIMD_LEVEL to CMake Options (default sse). add AVX levels (EZ_AVX_*) as seperate defines.  Set EZ_SIMD_IMPLEMENTATION respecting EZ_REQUESTED_SIMD_LEVEL.

TrevorCash committed:
Update Features_Platform with BUILDSYSTEM_REQUESTED_SIMD_LEVEL to set EZ_SIMD_IMPLEMENTATION.

TrevorCash committed:
set SIMD target_compile_options respecting EZ_REQUESTED_SIMD_LEVEL.

TrevorCash commented:
All The intrinsics are AVX (1).  So that should be about a 95% support rate according to steam.  I added EZ_REQUESTED_SIMD_LEVEL for control during the cmake process (with higher numbers generally meaning more SIMD capability).  Defaulting to the current SSE implementation.  I'd like to see Double precision support for Transformations.

C-Core commented:
@TrevorCash I opened a PR with the CMake setup how I envisioned it: https://github.com/ezEngine/ezEngine/pull/1731
This also allows us to go even lower than SSE 4.1 if needed since we also had complains about that sometimes.
I've added a stub of your SimdDouble class so you can rebase this PR on top of it once it is in. 

Note that ezSimdDouble should always be double precision no matter which SSE level is selected, we can fallback to FPU if you don't want to deal with implementing the 2 __m128d variant (which I can totally understand, I wouldn't want to do it either ;) ).

TrevorCash committed:
Merge remote-tracking branch 'origin/dev' into simd-double-avx

TrevorCash committed:
remove EZ_REQUESTED_SIMD_LEVEL

TrevorCash committed:
remove avx instructions from SSE directory.

TrevorCash committed:
populate SSEDouble_inl.h with generally 2 sse instructions for QuadDouble. update tests

TrevorCash committed:
update Features_Platform for explicit cases.  commit the SSEDouble_inl.h

TrevorCash commented:
for EZ_SIMD_IMPLEMENTATION = AVX, ezSimdDouble uses generally 1 AVX instruction per function. other classes use SSE implementation files unless AVX needed (ezSimdDouble, and future vec4d and similar types) . 

for EZ_SIMD_IMPLEMENTATION = SSE, ezSimdDouble uses generally 2 SSE instructions per function.

C-Core commented on Code/Engine/Foundation/Basics/AllDefinesOff.h:
```diff
@@ -92,15 +92,19 @@
 #define EZ_SIMD_IMPLEMENTATION_FPU 1
 #define EZ_SIMD_IMPLEMENTATION_SSE 2
 #define EZ_SIMD_IMPLEMENTATION_NEON 3
+#define EZ_SIMD_IMPLEMENTATION_AVX 4
```
Please revert these changes, we don't want to treat AVX as a separate simd implementation but as an advanced version of SSE (as discussed in my other PR).

C-Core commented on Code/Engine/Foundation/SimdMath/Implementation/SSE/SSEDouble_inl.h:
```diff
@@ -1,3 +1,393 @@
+
+  __m128i packedInt32 = _mm_set1_epi32(i);
+  m_v.xy = _mm_cvtepi32_pd(packedInt32);
+  m_v.zw = _mm_cvtepi32_pd(packedInt32);
```
m_v.zw = m_v.xy; no need to convert the same thing twice. This goes for all operations here. For simddouble we can assume that x,y,z and w all contain the same value.

TrevorCash replied:
duplicate commands should be removed in https://github.com/ezEngine/ezEngine/pull/1733

C-Core commented on Code/Engine/Foundation/SimdMath/Implementation/SSE/SSEVec4f_inl.h:
```diff
@@ -578,41 +578,25 @@ EZ_ALWAYS_INLINE ezSimdVec4f ezSimdVec4f::CrossRH(const ezSimdVec4f& v) const
 // static
 EZ_ALWAYS_INLINE ezSimdVec4f ezSimdVec4f::MulAdd(const ezSimdVec4f& a, const ezSimdVec4f& b, const ezSimdVec4f& c)
 {
-#if EZ_SSE_LEVEL >= EZ_SSE_AVX2
```
This should not be deleted.

TrevorCash replied:
if a AVX/AVXVec4f_inl.h file were created should the #if EZ_SSE_LEVEL >= EZ_SSE_AVX2 be removed from this file?  For EZ_MIN_REQUIRED_SSE_LEVEL = SSE4.1 the instructions in the ifdef should not evaluate anyway.

TrevorCash commented:
I implemented your suggestion for removing duplicate operations thanks for that.  

I Also Reverted the changes to SSE/SSEVec4f_inl.h (I removed the comments because the possibility of adding AVX/AVXVec4f_inl.h)

Removing #define EZ_SIMD_IMPLEMENTATION_AVX 4 really doesnt make sense to me - as it lives alongside NEON and SSE as another equal instrinic technology.

TrevorCash commented:
Sorry I closed the thread while I was re-organising things.  I re-opened the PR here: https://github.com/ezEngine/ezEngine/pull/1733. Sorry to change threads.
